pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -210,9 +210,15 @@ public void asyncAddData(T data, AddDataCallback callback, Object ctx){
             return;
         }
         singleThreadExecutorForWrite.execute(() -> {
+            int recordsCountBeforeAdd = dataArray.size();
             try {
                 internalAsyncAddData(data, callback, ctx);
             } catch (Exception e){
+                // Avoid missing callback, do failed callback when error occur before add data to the array.
+                int recordsCountAfter = dataArray.size();
+                if (recordsCountAfter == recordsCountBeforeAdd){
+                    callback.addFailed(new ManagedLedgerException.ManagedLedgerFencedException(e), ctx);
```

Comment:
> import static ManagedLedgerException to shorten expression.

The ManagedLedgerException created here uses the parameter `e`, so it cannot be set as a static variable.

> Can you please explain why did you choose the Fenced exception for this specific error which you don't know its nature?

No exception is thrown in the block of "try catch". If it does, it will be due to coding errors such as `data Serializer` or `metrics`.

Reply:
I'm not referring to static variable, but to add an import static statement, so you can only `new ManagedLedgerFencedException(` without the `ManagedLedgerException` prefix.

>No exception is thrown in the block of "try catch". If it does, it will be due to coding errors such as data Serializer or metrics.
I don't understand. You could have chosen RuntimeException or any other exception class, why did you specifically choose ManagedLedgerFencedException? 
@codelipenghui this sound reasonable to you?

Reply:
I agree, we should not throw ManagedLedgerException out of ManagedLedger.
IMO, we should use `PersistenceException` instead.